ENIGMAsystem/SHELL/Universal_System/Serialization/serialization.h

```diff
@@ -0,0 +1,817 @@
+}
+
+template <typename T, typename U>
+void internal_resize_buffer_for_map(std::vector<std::byte> &buffer, const std::map<T, U> &value) {
+  buffer.resize(buffer.size() +
+                value.size() * ((value.size()) ? (enigma_internal_sizeof(value.begin()->first) +
+                                                  enigma_internal_sizeof(value.begin()->second))
+                                               : 0) +
+                sizeof(std::size_t));
+}
+
+template <typename T>
+void internal_resize_buffer_for_complex(std::vector<std::byte> &buffer, const std::complex<T> &value) {
+  buffer.resize(
+      buffer.size() +
+      2 * enigma_internal_sizeof(value.real()));  // we don't need to store the size of the complex, it is always 1
+}
+
+template <typename T, typename = std::enable_if_t<has_byte_size_method_v<T>>>
+inline void internal_resize_buffer_using_byte_size(std::vector<std::byte> &buffer, const T &value) {
+  buffer.resize(buffer.size() + value.byte_size());
+}
+
+template <typename T>
+typename std::enable_if<std::is_same_v<var, std::decay_t<T>>>::type inline internal_resize_buffer_for_fn(
+    std::vector<std::byte> &buffer, T &&value) {
+  internal_resize_buffer_for_var(buffer, value);
+}
+
+template <typename T>
+typename std::enable_if<std::is_base_of_v<variant, std::decay_t<T>> && !std::is_same_v<var, std::decay_t<T>>>::
+    type inline internal_resize_buffer_for_fn(std::vector<std::byte> &buffer, T &&value) {
+  internal_resize_buffer_for_variant(buffer, value);
+}
+
+template <typename T>
+typename std::enable_if<std::is_same_v<std::string, std::decay_t<T>>>::type inline internal_resize_buffer_for_fn(
+    std::vector<std::byte> &buffer, T &&value) {
+  internal_resize_buffer_for_string(buffer, value);
+}
+
+template <typename T>
+typename std::enable_if<is_std_vector_v<std::decay_t<T>>>::type inline internal_resize_buffer_for_fn(
+    std::vector<std::byte> &buffer, T &&value) {
+  internal_resize_buffer_for_vector(buffer, value);
+}
+
+template <typename T>
+typename std::enable_if<is_std_map_v<std::decay_t<T>>>::type inline internal_resize_buffer_for_fn(
+    std::vector<std::byte> &buffer, T &&value) {
+  internal_resize_buffer_for_map(buffer, value);
+}
+
+template <typename T>
+typename std::enable_if<is_std_complex_v<std::decay_t<T>>>::type inline internal_resize_buffer_for_fn(
+    std::vector<std::byte> &buffer, T &&value) {
+  internal_resize_buffer_for_complex(buffer, value);
+}
+
+template <typename T>
+inline void internal_resize_buffer_for(std::vector<std::byte> &buffer, T &&value) {
+  if constexpr (has_internal_resize_buffer_for_fn_free_function<std::decay_t<T>>) {
+    internal_resize_buffer_for_fn(buffer, value);
+  } else if constexpr (has_byte_size_method_v<std::decay_t<T>>) {
+    internal_resize_buffer_using_byte_size(buffer, value);
+  } else {
+    internal_resize_buffer_for_value(buffer, value);
+  }
+}
+
+// A varint encoding version of the integer functions
```

Comment:
I should've done this in my project itself, but we can probably drop the VBR encoding functions which are commented out because these will probably never be used.